#!/usr/bin/env python3
"""
Cube Castle å…¨ç³»ç»Ÿé›†æˆæµ‹è¯•
"""

import asyncio
import json
import time
import requests
import grpc
import sys
import os
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed
import statistics

# é…ç½®æ—¥å¿—
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.insert(0, '/home/shangmeilin/cube-castle/python-ai')

try:
    import intelligence_pb2
    import intelligence_pb2_grpc
except ImportError:
    logger.error("æ— æ³•å¯¼å…¥gRPCæ¨¡å—ï¼Œè¯·ç¡®ä¿åœ¨python-aiç›®å½•ä¸‹è¿è¡Œ")
    sys.exit(1)

class CubeCastleIntegrationTest:
    """Cube Castle å…¨ç³»ç»Ÿé›†æˆæµ‹è¯•"""
    
    def __init__(self):
        self.base_url = "http://localhost:8080"
        self.ai_grpc_url = "localhost:50051"
        self.test_results = {
            "total": 0,
            "passed": 0,
            "failed": 0,
            "errors": []
        }
        
    def log_test_result(self, test_name: str, success: bool, message: str = ""):
        """è®°å½•æµ‹è¯•ç»“æœ"""
        self.test_results["total"] += 1
        if success:
            self.test_results["passed"] += 1
            logger.info(f"âœ… {test_name}: {message}")
        else:
            self.test_results["failed"] += 1
            self.test_results["errors"].append(f"{test_name}: {message}")
            logger.error(f"âŒ {test_name}: {message}")
    
    def test_health_endpoints(self):
        """æµ‹è¯•å¥åº·æ£€æŸ¥ç«¯ç‚¹"""
        logger.info("ğŸ” æµ‹è¯•å¥åº·æ£€æŸ¥ç«¯ç‚¹")
        
        try:
            response = requests.get(f"{self.base_url}/health", timeout=5)
            success = response.status_code == 200
            self.log_test_result(
                "å¥åº·æ£€æŸ¥ç«¯ç‚¹", 
                success, 
                f"çŠ¶æ€ç : {response.status_code}, å“åº”: {response.text}"
            )
        except Exception as e:
            self.log_test_result("å¥åº·æ£€æŸ¥ç«¯ç‚¹", False, f"è¯·æ±‚å¤±è´¥: {e}")
    
    def test_corehr_api(self):
        """æµ‹è¯•CoreHR API"""
        logger.info("ğŸ¢ æµ‹è¯•CoreHR API")
        
        # æµ‹è¯•å‘˜å·¥åˆ—è¡¨API
        try:
            response = requests.get(f"{self.base_url}/api/v1/corehr/employees", timeout=10)
            success = response.status_code == 200
            self.log_test_result(
                "å‘˜å·¥åˆ—è¡¨API", 
                success, 
                f"çŠ¶æ€ç : {response.status_code}"
            )
            
            if success:
                data = response.json()
                self.log_test_result(
                    "å‘˜å·¥åˆ—è¡¨æ•°æ®æ ¼å¼", 
                    "employees" in data,
                    f"å“åº”åŒ…å«employeeså­—æ®µ: {'employees' in data}"
                )
        except Exception as e:
            self.log_test_result("å‘˜å·¥åˆ—è¡¨API", False, f"è¯·æ±‚å¤±è´¥: {e}")
        
        # æµ‹è¯•ç»„ç»‡æ ‘API
        try:
            response = requests.get(f"{self.base_url}/api/v1/corehr/organizations/tree", timeout=10)
            success = response.status_code == 200
            self.log_test_result(
                "ç»„ç»‡æ ‘API", 
                success, 
                f"çŠ¶æ€ç : {response.status_code}"
            )
        except Exception as e:
            self.log_test_result("ç»„ç»‡æ ‘API", False, f"è¯·æ±‚å¤±è´¥: {e}")
    
    def test_employee_crud(self):
        """æµ‹è¯•å‘˜å·¥CRUDæ“ä½œ"""
        logger.info("ğŸ‘¤ æµ‹è¯•å‘˜å·¥CRUDæ“ä½œ")
        
        # åˆ›å»ºå‘˜å·¥
        employee_data = {
            "employee_number": f"TEST{int(time.time())}",
            "first_name": "æµ‹è¯•",
            "last_name": "å‘˜å·¥",
            "email": f"test{int(time.time())}@example.com",
            "status": "active"
        }
        
        try:
            # åˆ›å»ºå‘˜å·¥
            response = requests.post(
                f"{self.base_url}/api/v1/corehr/employees",
                json=employee_data,
                timeout=10
            )
            
            if response.status_code == 201:
                created_employee = response.json()
                employee_id = created_employee.get("id")
                
                self.log_test_result(
                    "åˆ›å»ºå‘˜å·¥", 
                    True, 
                    f"å‘˜å·¥ID: {employee_id}"
                )
                
                # è·å–åˆ›å»ºçš„å‘˜å·¥
                get_response = requests.get(
                    f"{self.base_url}/api/v1/corehr/employees",
                    timeout=10
                )
                
                if get_response.status_code == 200:
                    employees = get_response.json().get("employees", [])
                    found = any(emp.get("id") == employee_id for emp in employees)
                    self.log_test_result(
                        "è·å–åˆ›å»ºçš„å‘˜å·¥", 
                        found, 
                        f"åœ¨å‘˜å·¥åˆ—è¡¨ä¸­æ‰¾åˆ°æ–°åˆ›å»ºçš„å‘˜å·¥: {found}"
                    )
                else:
                    self.log_test_result("è·å–åˆ›å»ºçš„å‘˜å·¥", False, f"è·å–å‘˜å·¥åˆ—è¡¨å¤±è´¥: {get_response.status_code}")
            
            elif response.status_code == 409:
                self.log_test_result("åˆ›å»ºå‘˜å·¥", True, "å‘˜å·¥å·²å­˜åœ¨(409) - è¿™æ˜¯é¢„æœŸè¡Œä¸º")
            else:
                self.log_test_result("åˆ›å»ºå‘˜å·¥", False, f"åˆ›å»ºå¤±è´¥ï¼ŒçŠ¶æ€ç : {response.status_code}, å“åº”: {response.text}")
                
        except Exception as e:
            self.log_test_result("åˆ›å»ºå‘˜å·¥", False, f"è¯·æ±‚å¤±è´¥: {e}")
    
    def test_ai_service_grpc(self):
        """æµ‹è¯•AIæœåŠ¡gRPCæ¥å£"""
        logger.info("ğŸ¤– æµ‹è¯•AIæœåŠ¡gRPCæ¥å£")
        
        try:
            channel = grpc.insecure_channel(self.ai_grpc_url)
            stub = intelligence_pb2_grpc.IntelligenceServiceStub(channel)
            
            # æµ‹è¯•æ–‡æœ¬è§£é‡Š
            test_cases = [
                {
                    "text": "æ›´æ–°æˆ‘çš„ç”µè¯å·ç ä¸º13800138000",
                    "expected_intent": None,  # ä¸å¼ºåˆ¶è¦æ±‚ç‰¹å®šæ„å›¾
                    "description": "ç”µè¯å·ç æ›´æ–°"
                },
                {
                    "text": "è°æ˜¯æˆ‘çš„ç»ç†ï¼Ÿ",
                    "expected_intent": None,
                    "description": "æŸ¥è¯¢ç»ç†"
                },
                {
                    "text": "æŸ¥çœ‹æˆ‘çš„ä¸ªäººä¿¡æ¯",
                    "expected_intent": None,
                    "description": "æŸ¥çœ‹ä¸ªäººä¿¡æ¯"
                }
            ]
            
            for i, case in enumerate(test_cases):
                try:
                    request = intelligence_pb2.InterpretRequest()
                    request.user_text = case["text"]
                    request.session_id = f"integration-test-{i}"
                    
                    start_time = time.time()
                    response = stub.InterpretText(request, timeout=30)
                    end_time = time.time()
                    
                    response_time = (end_time - start_time) * 1000
                    
                    # éªŒè¯å“åº”
                    has_intent = hasattr(response, 'intent') and response.intent is not None
                    has_data = hasattr(response, 'structured_data_json') and response.structured_data_json is not None
                    
                    success = has_intent and has_data
                    self.log_test_result(
                        f"AIæ–‡æœ¬è§£é‡Š-{case['description']}", 
                        success, 
                        f"å“åº”æ—¶é—´: {response_time:.0f}ms, æ„å›¾: {response.intent if has_intent else 'None'}"
                    )
                    
                except grpc.RpcError as e:
                    self.log_test_result(
                        f"AIæ–‡æœ¬è§£é‡Š-{case['description']}", 
                        False, 
                        f"gRPCé”™è¯¯: {e.code()}, {e.details()}"
                    )
                except Exception as e:
                    self.log_test_result(
                        f"AIæ–‡æœ¬è§£é‡Š-{case['description']}", 
                        False, 
                        f"å¼‚å¸¸: {e}"
                    )
            
            channel.close()
            
        except Exception as e:
            self.log_test_result("AIæœåŠ¡gRPCè¿æ¥", False, f"è¿æ¥å¤±è´¥: {e}")
    
    def test_intelligence_gateway_api(self):
        """æµ‹è¯•Intelligence Gateway API"""
        logger.info("ğŸ§  æµ‹è¯•Intelligence Gateway API")
        
        test_data = {
            "query": "æˆ‘æƒ³æ›´æ–°æˆ‘çš„ç”µè¯å·ç ä¸º13900139000",
            "user_id": "11111111-1111-1111-1111-111111111111"
        }
        
        try:
            response = requests.post(
                f"{self.base_url}/api/v1/intelligence/interpret",
                json=test_data,
                timeout=30
            )
            
            success = response.status_code in [200, 500]  # å…è®¸500å› ä¸ºå¯èƒ½çš„AIæœåŠ¡é—®é¢˜
            
            if response.status_code == 200:
                data = response.json()
                has_intent = "intent" in data
                has_data = "structured_data_json" in data
                
                self.log_test_result(
                    "Intelligence Gateway API", 
                    has_intent and has_data, 
                    f"çŠ¶æ€ç : {response.status_code}, æ„å›¾: {data.get('intent', 'None')}"
                )
            else:
                self.log_test_result(
                    "Intelligence Gateway API", 
                    False, 
                    f"çŠ¶æ€ç : {response.status_code}, å“åº”: {response.text}"
                )
                
        except Exception as e:
            self.log_test_result("Intelligence Gateway API", False, f"è¯·æ±‚å¤±è´¥: {e}")
    
    def test_performance_benchmarks(self):
        """æµ‹è¯•æ€§èƒ½åŸºå‡†"""
        logger.info("âš¡ æµ‹è¯•æ€§èƒ½åŸºå‡†")
        
        # æµ‹è¯•APIå“åº”æ—¶é—´
        api_endpoints = [
            "/health",
            "/api/v1/corehr/employees",
            "/api/v1/corehr/organizations/tree"
        ]
        
        for endpoint in api_endpoints:
            response_times = []
            
            for i in range(5):  # æ¯ä¸ªç«¯ç‚¹æµ‹è¯•5æ¬¡
                try:
                    start_time = time.time()
                    response = requests.get(f"{self.base_url}{endpoint}", timeout=10)
                    end_time = time.time()
                    
                    if response.status_code == 200:
                        response_time = (end_time - start_time) * 1000
                        response_times.append(response_time)
                except Exception:
                    pass
            
            if response_times:
                avg_time = statistics.mean(response_times)
                max_time = max(response_times)
                min_time = min(response_times)
                
                # APIå“åº”æ—¶é—´åŸºå‡†ï¼šå¹³å‡ < 1000msï¼Œæœ€å¤§ < 3000ms
                performance_ok = avg_time < 1000 and max_time < 3000
                
                self.log_test_result(
                    f"APIæ€§èƒ½-{endpoint}", 
                    performance_ok, 
                    f"å¹³å‡: {avg_time:.0f}ms, æœ€å¤§: {max_time:.0f}ms, æœ€å°: {min_time:.0f}ms"
                )
            else:
                self.log_test_result(f"APIæ€§èƒ½-{endpoint}", False, "æ— æœ‰æ•ˆå“åº”æ—¶é—´æ•°æ®")
    
    def test_concurrent_load(self):
        """æµ‹è¯•å¹¶å‘è´Ÿè½½"""
        logger.info("ğŸš€ æµ‹è¯•å¹¶å‘è´Ÿè½½")
        
        def make_request():
            try:
                response = requests.get(f"{self.base_url}/health", timeout=5)
                return response.status_code == 200
            except:
                return False
        
        # å¹¶å‘æµ‹è¯•ï¼š10ä¸ªå¹¶å‘è¯·æ±‚
        with ThreadPoolExecutor(max_workers=10) as executor:
            futures = [executor.submit(make_request) for _ in range(10)]
            results = [future.result() for future in as_completed(futures)]
        
        success_count = sum(results)
        success_rate = success_count / len(results) * 100
        
        # è¦æ±‚æˆåŠŸç‡ >= 80%
        load_test_ok = success_rate >= 80
        
        self.log_test_result(
            "å¹¶å‘è´Ÿè½½æµ‹è¯•", 
            load_test_ok, 
            f"æˆåŠŸç‡: {success_rate:.1f}% ({success_count}/{len(results)})"
        )
    
    def test_data_consistency(self):
        """æµ‹è¯•æ•°æ®ä¸€è‡´æ€§"""
        logger.info("ğŸ”’ æµ‹è¯•æ•°æ®ä¸€è‡´æ€§")
        
        try:
            # è·å–å‘˜å·¥åˆ—è¡¨ä¸¤æ¬¡ï¼ŒéªŒè¯æ•°æ®ä¸€è‡´æ€§
            response1 = requests.get(f"{self.base_url}/api/v1/corehr/employees", timeout=10)
            time.sleep(1)  # ç­‰å¾…1ç§’
            response2 = requests.get(f"{self.base_url}/api/v1/corehr/employees", timeout=10)
            
            if response1.status_code == 200 and response2.status_code == 200:
                data1 = response1.json()
                data2 = response2.json()
                
                # æ¯”è¾ƒå‘˜å·¥æ•°é‡ï¼ˆåœ¨æ²¡æœ‰å¹¶å‘ä¿®æ”¹çš„æƒ…å†µä¸‹åº”è¯¥ç›¸åŒï¼‰
                count1 = len(data1.get("employees", []))
                count2 = len(data2.get("employees", []))
                
                consistency_ok = count1 == count2
                
                self.log_test_result(
                    "æ•°æ®ä¸€è‡´æ€§", 
                    consistency_ok, 
                    f"ç¬¬ä¸€æ¬¡æŸ¥è¯¢: {count1}æ¡, ç¬¬äºŒæ¬¡æŸ¥è¯¢: {count2}æ¡"
                )
            else:
                self.log_test_result("æ•°æ®ä¸€è‡´æ€§", False, "æ— æ³•è·å–æ•°æ®è¿›è¡Œä¸€è‡´æ€§æ£€æŸ¥")
                
        except Exception as e:
            self.log_test_result("æ•°æ®ä¸€è‡´æ€§", False, f"æµ‹è¯•å¤±è´¥: {e}")
    
    def run_all_tests(self):
        """è¿è¡Œæ‰€æœ‰é›†æˆæµ‹è¯•"""
        logger.info("ğŸ° å¼€å§‹ Cube Castle å…¨ç³»ç»Ÿé›†æˆæµ‹è¯•")
        logger.info("=" * 60)
        
        start_time = time.time()
        
        # æŒ‰é€»è¾‘é¡ºåºæ‰§è¡Œæµ‹è¯•
        self.test_health_endpoints()
        self.test_corehr_api()
        self.test_employee_crud()
        self.test_ai_service_grpc()
        self.test_intelligence_gateway_api()
        self.test_performance_benchmarks()
        self.test_concurrent_load()
        self.test_data_consistency()
        
        end_time = time.time()
        total_time = end_time - start_time
        
        # è¾“å‡ºæµ‹è¯•æ€»ç»“
        logger.info("=" * 60)
        logger.info("ğŸ° Cube Castle å…¨ç³»ç»Ÿé›†æˆæµ‹è¯•å®Œæˆ")
        logger.info(f"æ€»æµ‹è¯•æ—¶é—´: {total_time:.2f}ç§’")
        logger.info(f"æ€»æµ‹è¯•æ•°é‡: {self.test_results['total']}")
        logger.info(f"âœ… é€šè¿‡: {self.test_results['passed']}")
        logger.info(f"âŒ å¤±è´¥: {self.test_results['failed']}")
        
        if self.test_results["failed"] > 0:
            logger.info("\nå¤±è´¥çš„æµ‹è¯•:")
            for error in self.test_results["errors"]:
                logger.info(f"  - {error}")
        
        success_rate = (self.test_results["passed"] / self.test_results["total"]) * 100
        logger.info(f"æˆåŠŸç‡: {success_rate:.1f}%")
        
        if success_rate >= 90:
            logger.info("ğŸ‰ ç³»ç»Ÿæ•´ä½“çŠ¶æ€: ä¼˜ç§€")
        elif success_rate >= 75:
            logger.info("âœ… ç³»ç»Ÿæ•´ä½“çŠ¶æ€: è‰¯å¥½")
        elif success_rate >= 50:
            logger.info("âš ï¸  ç³»ç»Ÿæ•´ä½“çŠ¶æ€: éœ€è¦æ”¹è¿›")
        else:
            logger.info("âŒ ç³»ç»Ÿæ•´ä½“çŠ¶æ€: å­˜åœ¨ä¸¥é‡é—®é¢˜")
        
        logger.info("=" * 60)
        
        return self.test_results

if __name__ == "__main__":
    test_runner = CubeCastleIntegrationTest()
    results = test_runner.run_all_tests()
    
    # æ ¹æ®æµ‹è¯•ç»“æœè®¾ç½®é€€å‡ºä»£ç 
    exit_code = 0 if results["failed"] == 0 else 1
    sys.exit(exit_code)